### ISSUE TITLE
MacOS compilation

### ISSUE THREAD

**Issue by Valeria** [offa/plug#9]
Hello.

I'm trying to compile this on Mac (I know this is meant for Linux, but I believe at least I could get the application itself compiled even if it later has issues connecting to the amp).

Even though I have installed qt5, libusb-1.0 and googletest I'm getting fatal errors about libusb-1.0 not being found.

Here's the output of cmake:

>  cmake ..
>
>-- The C compiler identification is AppleClang 13.0.0.13000029
>-- The CXX compiler identification is AppleClang 13.0.0.13000029
>-- Detecting C compiler ABI info
>-- Detecting C compiler ABI info - done
>-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
>-- Detecting C compile features
>-- Detecting C compile features - done
>-- Detecting CXX compiler ABI info
>-- Detecting CXX compiler ABI info - done
>-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
>-- Detecting CXX compile features
>-- Detecting CXX compile features - done
>-- ~~~ plug v1.4.3 ~~~
>-- Unit Tests : ON
>-- Integrationtests : OFF
>-- Coverage : OFF
>-- LTO : OFF
>-- ASan : OFF
>-- UBSan : OFF
>-- Build Type : None
>-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2")
>-- Found libusb-1.0: /usr/local/Cellar/libusb/1.0.26/lib/libusb-1.0.dylib
>-- Looking for pthread.h
>-- Looking for pthread.h - found
>-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
>-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
>-- Found Threads: TRUE
>-- Found GTest: /usr/local/include
>-- Configuring done
>-- Generating done
>-- Build files have been written to: /Users/valeria/Downloads/plug-master/build

It seems to find libusb (installed via Homebrew) yet on the make part it doesn't seem to link it:

`fatal error: 'libusb-1.0/libusb.h' file not found`

I've tried manually editing all files in /src/com which have a libusb #include and swapping them to point to my /usr/local/Cellar/libusb/1.0.26/ libusb.h file yet it seems than it still shows an error when compiling the LibUsbCompat.cpp file, even though it doesn't seem to contain any libusb.h includes.

I'm also getting a ton of errors about old-style-casting and unsigned-conversions, which I've ignored using make -i

Am I missing something? Thank you.

**Comment by j824h** [comment]
I had to add 
`include_directories(/usr/local/include)`
in `src/com/CMakeLists.txt`, so that CMake can find libusb-1.0 installed by Homebrew.

Also for the root `CMakeLists.txt`, I found disabling `-pedantic`, `-pedantic-errors`, `-Wold-style-cast` help.

**Comment by offa** [comment]
What does your `libusb-1.0_INCLUDE_DIR` show? 

Regarding warnings: Did compilation fail otherwise? What errors are shown? CI builds include Clang too, but only on Linux.

**Comment by j824h** [comment]
### Include errors
`-- Found libusb-1.0: /usr/local/Cellar/libusb/1.0.26/lib/libusb-1.0.dylib`
I see that `libusb-1.0_INCLUDE_DIR` is `/usr/local/include`.

Still, `make` would complain.
> /Users/j824h/Documents/plug/src/com/UsbComm.cpp:25:10: fatal error: 'libusb-1.0/libusb.h' file not found

### Syntax errors
After `include_directories(/usr/local/include)`, the errors come from the header.

> In file included from /Users/j824h/Documents/plug/src/com/UsbComm.cpp:25:
> /usr/local/include/libusb-1.0/libusb.h:168:15: error: use of old-style cast [-Werror,-Wold-style-cast]
>         _tmp.b8[1] = (uint8_t) (x >> 8);
>                      ^         ~~~~~~~~
> /usr/local/include/libusb-1.0/libusb.h:169:15: error: use of old-style cast [-Werror,-Wold-style-cast]
>         _tmp.b8[0] = (uint8_t) (x & 0xff);
>                      ^         ~~~~~~~~~~
> /usr/local/include/libusb-1.0/libusb.h:789:31: error: zero size arrays are an extension [-Werror,-Wzero-length-array]
>         uint8_t  dev_capability_data[ZERO_SIZED_ARRAY];
... and so on.

Clang correctly cast errors for the non-C++ syntax in libusb.h, according to the flags provided in `CMakeLists.txt`
In the other way round, I started to wonder why CI could build.

**Comment by offa** [comment]
> Still, make would complain.

But `/usr/local/include` contains `libusb-1.0/libusb.h`!? Well … than there's something broken [here](https://github.com/offa/plug/blob/master/cmake/Findlibusb-1.0.cmake).

> Clang correctly cast errors for the non-C++ syntax in libusb.h, according to the flags provided in CMakeLists.txt
In the other way round, I started to wonder why CI could build.

Yes and no, the reported issues come from an external library (libusb) and those are treated as system headers, which aren't affected by warnings etc..

Including through `include_directories()` wont do this, but rather treat the headers as usual. You can test this (_not_ a solution!):

```cmake
include_directories(SYSTEM/usr/local/include)
```

I'm not sure whether this works on Mac Clang though.


To sum it up: `Findlibusb-1.0.cmake` isn't working correctly here (--> include not found).

**Comment by offa** [comment]
Can you try the `237-libusb_missing_header` branch please? :facepalm:

**Comment by j824h** [comment]
> > Still, make would complain.
> 
> But `/usr/local/include` contains `libusb-1.0/libusb.h`!? Well … than there's something broken [here](https://github.com/offa/plug/blob/master/cmake/Findlibusb-1.0.cmake).
> 
> Relying on PkgConfig not supported on Mac?
> 
> > Clang correctly cast errors for the non-C++ syntax in libusb.h, according to the flags provided in CMakeLists.txt
> > In the other way round, I started to wonder why CI could build.
> 
> Yes and no, the reported issues come from an external library (libusb) and those are treated as system headers, which aren't affected by warnings etc..
> 
> Including through `include_directories()` wont do this, but rather treat the headers as usual. You can test this (_not_ a solution!):
> 
> ```cmake
> include_directories(SYSTEM /usr/local/include)
> ```
> 
> I'm not sure whether this works on Mac Clang though.
> 
> To sum it up: `Findlibusb-1.0.cmake` isn't working correctly here (--> include not found).

Oh, thanks for the clarification! (I got much left to know about this CMake...)
```include_directories(SYSTEM /usr/local/include)```
For now, adding this is all I need to build.

> Can you try the `237-libusb_missing_header` branch please? 🤦

I will keep tracking this!
Now `make` moves on to other targets referring to `libusb.h` and gives errors there.

**Comment by offa** [comment]
Thanks, so I can merge the branch. What are the other targets / errors?

**Comment by j824h** [comment]
- plug-communication-usb
    - UsbContext.cpp
    - UsbException.cpp
    - UsbDevice.cpp
- plug-updater
    - MustangUpdater.cpp

All of those, excluding the tests.

**Comment by offa** [comment]
Seems these targets need some more work.

**Comment by j824h** [comment]
Thanks!
Adding `target_link_libraries` worked with the right target specified. (https://github.com/offa/plug/compare/237-libusb...j824h:plug:patch-1)

**Comment by offa** [comment]
Great! Since you have the change already, could you submit a PR please?

**Comment by offa** [comment]
Changes merged into `master`. Are there any build issues left?

**Comment by j824h** [comment]
In a Homebrew environment, there's a subtlety, that `Qt5_DIR=/usr/local/opt/qt5/lib/cmake/Qt5` should be passed to CMake.
I do think the blame is on the package manager's choice not to install Qt in a default path.
Otherwise, https://github.com/offa/plug/commit/8d93d21ab26107442a144851e5074871b4dd3478 is working great with `libusb-1.0` and `googletest` installed by Homebrew on MacOS.

**Comment by offa** [comment]
I'm going to close here since compilation works now. Feel free to reopen if there's an issue left. Thanks :+1:
